lib/s3routes/routesUtils.ts

```diff
@@ -441,23 +447,42 @@ function retrieveData(
                 });
                 // errors on server side with readable stream
                 readable.on('error', err => {
-                    log.error('error piping data from source');
+                    log.error('error piping data from source', { error: err });
                     _destroyResponse();
                     return cbOnce(err);
                 });
+                // At this point, we assume that the data backend is available.
+                // We write the http headers before piping the data to the
+                // response object.
+                // Any subsequent error leads to destror deletion.
+                // todo: need to check if headers were already set
+                if (!response.headersSent) {
+                    response.writeHead(httpCodeOnSucess);
+                }
+                // TODO ARSN-474: can this be moved above, so we
+                // do not need to call _destroyReadable in the error case
+                // in the callback of data.get()? The 'close' event is not
+                // called if 'end' is.
                 currentStream = readable;
                 return readable.pipe(response, { end: false });
             }), err => {
-            currentStream = null;
-            if (err) {
-                log.debug('abort response due to error', {
-                    // @ts-expect-error
-                    error: err.code, errMsg: err.message });
-            }
-            // call end for all cases (error/success) per node.js docs
-            // recommendation
-            response.end();
-        },
+                currentStream = null;
+                if (err) {
+                    log.error('abort response due to error', {
```

Comment:
`abort` is only in case the header was already sent, and we don't send an errorResponse. When we reach line 475, we are just sending a "regular" error.

Reply:
The logs were moved out of the `err` case, and consistently log once based on the error. Also we don't call `end` if the response was destroyed.